File: src/no/omb/bigboat/io/CsvReader.java
package no.omb.bigboat.io;

import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import no.omb.bigboat.BigBoat;
import no.omb.bigboat.data.Boat;
import no.omb.bigboat.data.RaceData;
import no.omb.bigboat.data.RaceEntry;
import no.omb.bigboat.data.SeriesEntry;
import no.omb.bigboat.data.RaceData.DataFileType;
import au.com.bytecode.opencsv.CSVReader;

public class CsvReader implements DataReader {

	private static CsvReader instance = new CsvReader();

	private RaceData race;

	@Override
	public List<RaceEntry> parseDataFile(RaceData race) {
		this.race = race;
		List<String[]> entries = readDataFile(getFileName(race));
		return parseEntries(entries);
	}

	private List<String[]> readDataFile(String fileName) {
		CSVReader reader;
		try {
			FileInputStream is = new FileInputStream(fileName);
			reader = new CSVReader(new InputStreamReader(is, "UTF-8"), BigBoat.SEP);
			List<String[]> entries = reader.readAll();
			reader.close();
			return entries;
		} catch (FileNotFoundException e) {
			e.printStackTrace();
		} catch (IOException e) {
			e.printStackTrace();
		}
		return null;
	}

	private List<RaceEntry> parseEntries(List<String[]> entries) {
		List<RaceEntry> raceEntries = new ArrayList<>();
		int classStartIndex = findClassStart(entries, 0);
		while (classStartIndex != -1) {
			int classSize = findClassSize(entries, classStartIndex);
			for (int ix = classStartIndex; ix < classStartIndex+classSize; ++ix) {
				RaceEntry newEntry = parseEntry(entries.get(ix), classSize);
				addOrUpdateRaceEntry(raceEntries, newEntry);
			}
			classStartIndex = findClassStart(entries, classStartIndex + classSize);
		}
		Collections.sort(raceEntries);
		return raceEntries;
	}

	private void addOrUpdateRaceEntry(List<RaceEntry> raceEntries, RaceEntry newEntry) {
		for (RaceEntry entry : raceEntries) {
			if (entry.getBoat().equals(newEntry.getBoat())) {
				if (entry.getScore() > newEntry.getScore()) {
					entry.applyValues(newEntry);
				}
				return;
			}
		}
		raceEntries.add(newEntry);
	}

	private RaceEntry parseEntry(String[] col, int classSize) {
		RaceEntry entry = new RaceEntry();
		try {
			if (race.getDataFileType() == DataFileType.SEILMAG_CSV4) {
				entry.setPlaceNo(Integer.parseInt(col[0]));
				entry.setBoat(new Boat(col[1], col[2], col[3], col[4].split(" \\(")[0], col[4].split(" \\(")[1]));
			}
			else if (race.getDataFileType() == DataFileType.SEILMAG_CSV3) {
				entry.setPlaceNo(Integer.parseInt(col[0]));
				entry.setBoat(new Boat(col[1], col[2], col[3], col[8].split(" \\(")[0], col[8].split(" \\(")[1]));
			}
			else if (race.getDataFileType() == DataFileType.SEILMAG_CSV2) {
				entry.setPlaceNo(Integer.parseInt(col[0]));
				entry.setBoat(new Boat(col[1], col[2], col[3], col[9].split(" \\(")[0], col[9].split(" \\(")[1]));
			}
			else if (race.getDataFileType() == DataFileType.M2S_CSV) {
				entry.setPlaceNo(Integer.parseInt(col[0]));
				entry.setBoat(new Boat(col[1], col[2], col[3], col[4], col[5]));
			}
			else if (race.getDataFileType() == DataFileType.M2S_CSV2) {
				entry.setPlaceNo(Integer.parseInt(col[0]));
				entry.setBoat(new Boat(col[1], col[2], col[5], col[3], col[4]));
			}
			else if (race.getDataFileType() == DataFileType.M2S_CSV3) {
				entry.setPlaceNo(Integer.parseInt(col[0]));
				entry.setBoat(new Boat(col[1], col[5], col[4], col[2], col[3]));
			}
			else if (race.getDataFileType() == DataFileType.ULLERN_CSV) {
				entry.setPlaceNo(Integer.parseInt(col[0]));
				entry.setBoat(new Boat(col[3] + "-" + col[4], col[6], col[5], col[1], col[2]));
			}
			else {
				entry.setPlaceNo(Integer.parseInt(col[0]));
				entry.setBoat(new Boat(col[1], col[2], col[5], col[3], col[4]));
			}
		}
		catch (ArrayIndexOutOfBoundsException e) {
			System.out.println("Error in line: " + col);
		}
		entry.setClassSize(classSize);
		entry.setScore(entry.getPlaceNo() * SeriesEntry.SCORE_FACTOR * 1.0 / classSize);
		return entry;
	}

	private int findClassStart(List<String[]> entries, int startIndex) {
		for (int ix = startIndex; ix < entries.size(); ++ix) {
			try {
				Integer.parseInt(entries.get(ix)[0]);
				if (entries.get(ix).length <= 1) {
					continue;
				}
				return ix;
			}
			catch (NumberFormatException e) {
				// Do nothing
			}
		}
		return -1;
	}

	private int findClassSize(List<String[]> entries, int classStartIndex) {
		for (int ix = classStartIndex+1; ix < entries.size(); ++ix) {
			try {
				Integer.parseInt(entries.get(ix)[0]);
				if (entries.get(ix).length <= 1) {
					return ix - classStartIndex;
				}
			}
			catch (NumberFormatException e) {
				return ix - classStartIndex;
			}
		}
		return entries.size() - classStartIndex;
	}

	private String getFileName(RaceData race) {
		return BigBoat.DATA + "/" + race.getDataFileName() + ".csv";
	}

	public static CsvReader getInstance() {
		return instance;
	}

}

File: src/main/kotlin/no/omb/bigboat/io/CsvReader.kt
package no.omb.bigboat.io

import au.com.bytecode.opencsv.CSVReader
import no.omb.bigboat.BigBoat
import no.omb.bigboat.data.Boat
import no.omb.bigboat.data.RaceData
import no.omb.bigboat.data.RaceData.DataFileType
import no.omb.bigboat.data.RaceEntry
import no.omb.bigboat.data.SeriesEntry
import java.io.*
import java.util.*

class CsvReader : DataReader {
    private var race: RaceData? = null
    override fun parseDataFile(race: RaceData): List<RaceEntry> {
        this.race = race
        val entries = readDataFile(getFileName(race))
        return parseEntries(entries)
    }

    private fun readDataFile(fileName: String): List<Array<String>> {
        val reader: CSVReader
        try {
            val `is` = FileInputStream(fileName)
            reader = CSVReader(InputStreamReader(`is`, "UTF-8"), BigBoat.SEP)
            val entries = reader.readAll()
            reader.close()
            return entries
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return emptyList()
    }

    private fun parseEntries(entries: List<Array<String>>): List<RaceEntry> {
        val raceEntries: MutableList<RaceEntry> = ArrayList()
        var classStartIndex = findClassStart(entries, 0)
        while (classStartIndex != -1) {
            val classSize = findClassSize(entries, classStartIndex)
            for (ix in classStartIndex until classStartIndex + classSize) {
                val newEntry = parseEntry(entries[ix], classSize)
                addOrUpdateRaceEntry(raceEntries, newEntry)
            }
            classStartIndex = findClassStart(entries, classStartIndex + classSize)
        }
        raceEntries.sort()
        return raceEntries
    }

    private fun addOrUpdateRaceEntry(raceEntries: MutableList<RaceEntry>, newEntry: RaceEntry) {
        for (entry in raceEntries) {
            if (entry.boat == newEntry.boat) {
                if (entry.score > newEntry.score) {
                    entry.applyValues(newEntry)
                }
                return
            }
        }
        raceEntries.add(newEntry)
    }

    private fun parseEntry(col: Array<String>, classSize: Int): RaceEntry {
        val entry = RaceEntry()
        try {
            when (race!!.dataFileType) {
                DataFileType.SEILMAG_CSV4 -> {
                    entry.placeNo = col[0].toInt()
                    entry.boat = Boat(
                        col[1],
                        col[2],
                        col[3],
                        col[4].split(" \\(".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0],
                        col[4].split(" \\(".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[1]
                    )
                }

                DataFileType.SEILMAG_CSV3 -> {
                    entry.placeNo = col[0].toInt()
                    entry.boat = Boat(
                        col[1],
                        col[2],
                        col[3],
                        col[8].split(" \\(".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0],
                        col[8].split(" \\(".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[1]
                    )
                }

                DataFileType.SEILMAG_CSV2 -> {
                    entry.placeNo = col[0].toInt()
                    entry.boat = Boat(
                        col[1],
                        col[2],
                        col[3],
                        col[9].split(" \\(".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0],
                        col[9].split(" \\(".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[1]
                    )
                }

                // 0: Place, 1: Sail no, 2: Boat name, 3: Boat type, 4: Skipper, 5: Club
                DataFileType.M2S_CSV -> {
                    entry.placeNo = col[0].toInt()
                    entry.boat = Boat(col[1], col[2], col[3], col[4], col[5])
                }

                // 0: Place, 1: Sail no, 2: Boat name, 3: Skipper, 4: Club, 5: Boat type
                DataFileType.M2S_CSV2 -> {
                    entry.placeNo = col[0].toInt()
                    entry.boat = Boat(col[1], col[2], col[5], col[3], col[4])
                }

                // 0: Place, 1: Sail no, 2: Skipper, 3: Club, 4: Boat type, 5: Boat name
                DataFileType.M2S_CSV3 -> {
                    entry.placeNo = col[0].toInt()
                    entry.boat = Boat(col[1], col[5], col[4], col[2], col[3])
                }

                // 0: Place, 1: Sail no, 2: Skipper, 3: Boat name, 4: Club, 5: Boat type
                DataFileType.M2S_CSV4 -> {
                    entry.placeNo = col[0].toInt()
                    entry.boat = Boat(col[1], col[3], col[5], col[2], col[4])
                }

                // 0: Place, 1: Sail no, 2: Skipper, 3: Club, 4: Boat name, 5: Boat type
                DataFileType.M2S_CSV6 -> {
                    entry.placeNo = col[0].toInt()
                    entry.boat = Boat(col[1], col[4], col[5], col[2], col[3])
                }

                DataFileType.ULLERN_CSV -> {
                    entry.placeNo = col[0].toInt()
                    entry.boat = Boat(col[3] + "-" + col[4], col[6], col[5], col[1], col[2])
                }

                // 0: Place, 1: Sail no, 2: Boat type, 3: Boat name, 4: Skipper, 5: Club
                DataFileType.DROBAK_CSV -> {
                    entry.placeNo = col[0].toInt()
                    entry.boat = Boat(col[1], col[3], col[2], col[4], col[5])
                }

                else -> {
                    entry.placeNo = col[0].toInt()
                    entry.boat = Boat(col[1], col[2], col[5], col[3], col[4])
                }
            }
        } catch (e: ArrayIndexOutOfBoundsException) {
            println("Error in line: $col")
        }
        entry.classSize = classSize
        entry.score = entry.placeNo * SeriesEntry.SCORE_FACTOR * 1.0 / classSize
        return entry
    }

    private fun findClassStart(entries: List<Array<String>>, startIndex: Int): Int {
        for (ix in startIndex until entries.size) {
            try {
                entries[ix][0].toInt()
                if (entries[ix].size <= 1) {
                    continue
                }
                return ix
            } catch (e: NumberFormatException) {
                // Do nothing
            }
        }
        return -1
    }

    private fun findClassSize(entries: List<Array<String>>, classStartIndex: Int): Int {
        for (ix in classStartIndex + 1 until entries.size) {
            try {
                entries[ix][0].toInt()
                if (entries[ix].size <= 1) {
                    return ix - classStartIndex
                }
            } catch (e: NumberFormatException) {
                return ix - classStartIndex
            }
        }
        return entries.size - classStartIndex
    }

    private fun getFileName(race: RaceData): String {
        return BigBoat.DATA + "/" + race.dataFileName + ".csv"
    }

    companion object {
        val instance = CsvReader()
    }
}
